### DISCUSSION TITLE
Branches in magick scripts

### DISCUSSION THREAD

**Discussion by Tillerino** [ImageMagick/ImageMagick#7544] [Help] [answered]
### ImageMagick version

7.1.1-36

### Operating system, version and so on

NixOS

### Description

I am trying to batch-process a lot of old B&W photos, which on occasion, can be too much to dark.

I have found that `-auto-gamma` does a fine job with brightening those photos which are too dark, but when applied to bright photos, which e.g. feature a lot of clouds, it destroys the picture.

So what I would like to do here is _only_ apply the `-auto-gamma` flag _if_ it would brighten the photo. I have seen that I could measure the average intensity with `identify`. I wonder if I can somehow get a condition into a magick script such that the `-auto-gamma` flag is only executed if the average intensity is below a certain threshold.

**Comment by snibgo** [comment, marked as answer]
Magick scripts (`magick ... -script`) don't have a facility for conditionals `if ... then ...` or loops `while ... do ...`.

Instead, these can be done in shell (bash, BAT, etc) scripts. For example, I might have a `magick` command like this (Windows syntax):
```
magick %INFILE% %PROCESS% %OUTFILE%
```
... where `INFILE` etc are environment variables, and PROCESS might contain `-auto-gamma`, depending on a previous command.

Another possibility is for a single `magick` command to clone the image, then `-auto-gamma` the clone, then delete either the original image or the clone, depending on which is the darkest.
```
magick in.png ( +clone -auto-gamma ) -delete "%[fx:u.mean<v.mean?0:1]" out.png
```
Here, `u.mean` is the mean pixel value of `in.png`, and `v.mean` is the mean pixel value of the auto-gamma version. This isn't really "lightness", so a better version could convert the images to Lab, and use the mean value of the L channel.

**Reply by Tillerino** [reply]
Mh, I don't want to use a shell script, as this is part of a larger system and I want it to compose nicely.

The clone/delete trick should do nicely, though! As these are B&W photos, this will absolute correspond to the brighter image, no?

**Comment by snibgo** [comment]
> As these are B&W photos, this will absolute correspond to the brighter image, no?

Yes.

**Reply by fmw42** [reply]
Have you tried negating the image then -auto-gamma then negating the result?  Does that work for the bright ones?

**Reply by Tillerino** [reply]
It's not so much that it's a technical issue with the bright images. With my material, dark images are usually underexposed whereas bright images just feature a lot of cloudy sky, but are fine. I'm not trying to win any prizes with the results. These are old family photos, which I just want to make _viewable_.

Interesting idea, though. I'll check it out. Thanks!

**Comment by fmw42** [comment]
Can you post a few of the bright cloudy images so I can see if there is another better approach?
